Faucard patch1

BDrescher opened the pull request:
Hey all,
as the current faucard payment occured errors in relation to the misuse of floating points precision, I've today tried to account all occurences of those and patch them. Additionally I have updated the log generation used for the invoice to be able to skip errors manually and made it a little safer.

@mgmax @patkan I would be happy if you got the time to review my changes and maybe point at code passages which are not yet safe enough.

BDrescher committed:
Changed the to be payed amount to Decimal wherever possible. Using floating point safe conversion to cents

BDrescher committed:
Updated logmodule to use and Process Decimals, updated check_last_transaction to use EURCENTS / Decimals

BDrescher committed:
ADD Support to acknowledge missing bookings. Fixed uncertainties

include microseconds 999999 to enddate
use first and last booking to sum up kassenbuch

Simplified SQL command to query Kassenbuch entries

mgmax commented on FabLabKasse/faucardPayment/FAUcardPaymentThread.py:
```diff
@@ -78,7 +79,8 @@ def __init__(self, dialog, amount, thread=QtCore.QThread.currentThread()):
         self.old_balance = 0
         self.new_balance = 0
-        self.amount = int(amount * 100)
+        self.amount = amount
```
Does amount have to be float? Passing it as Decimal would be cleaner.

mgmax replied:
Changing the meaning of `amount` from cents to euros is dangerous. Maybe just remove the old `amount` variable and only use amount_cents?

BDrescher replied:
Amount has carefully been replaced with amount_cents. This variable is used throughout the whole payment procedure and amount is only used as a Decimal backup. You are definetely right in the point that we want a Decimal and I just not to enforce the usage by float_to_decimal.
At the moment i expect the amount to already be decimal (the prior conversion in faucard.py was removed)

mgmax commented on FabLabKasse/faucardPayment/MagPosLog.py:
```diff
@@ -34,13 +34,13 @@ def __init__(self, amount, cur, con):
 
         # Create table if it does not exist
-        self.cur.execute("CREATE TABLE IF NOT EXISTS MagPosLog(id INTEGER PRIMARY KEY AUTOINCREMENT, datum, cardnumber INT, amount FLOAT, oldbalance INT, newbalance INT, timestamp_payed, status INT, info INT, payed)")
+        self.cur.execute("CREATE TABLE IF NOT EXISTS MagPosLog(id INTEGER PRIMARY KEY AUTOINCREMENT, datum, cardnumber INT, amount TEXT, oldbalance INT, newbalance INT, timestamp_payed, status INT, info INT, payed)")
```
SQLite cannot (?) change the value of a column, and conversion will be quite annoying. I think we should rather accept storing the values as float and load them with float_to_decimal. As long as you don't go above 1e8 euros or so, there is no problem with float as long as you round correctly.

If you really do want to convert the database, then we could just switch to INT and store it as cents.

BDrescher replied:
The problem is that the database does not store it with unlimited precision as far as i noticed. The amount 2.04999 were therefore cut to be 2.04 ...
The conversion was already done this evening (SQLITE can do that on its own) but not yet pushed to Kasse. I verified the conversion by summing the columns and comparing side by side and have not yet seen any difference in data so far.

The column type TEXT was choosen to copy the structure of kassenbuch, where "betrag" is of type TEXT.

mgmax replied:
okay, that sounds convincing.

mgmax commented on FabLabKasse/faucardPayment/faucard.py:
```diff
@@ -30,7 +30,7 @@ def __init__(self, parent, amount):
         assert isinstance(amount, (Decimal, float)), "PayupFAUCard: Amount to pay not Decimal or float"
```
this line should be changed, as we only want to accept Decimal (?)

BDrescher replied:
You are correct. Thanks for the notice. As already mentioned this should be happening in FAUcardPaymentThread aswell.

mgmax commented on FabLabKasse/faucardPayment/generateLog.py:
```diff
@@ -126,7 +144,7 @@ def verify_sum(curKb, start, end, magposSum):
 
             # Determine corresponding rechnung
-            curKb.execute("SELECT rechnung FROM buchung WHERE konto = 'FAUKarte' AND (betrag = ? OR betrag = ?) AND datum BETWEEN ? AND ? ",(unicode(row[3]), "{:.3f}".format(row[3]), timestamp, timestamp + timedelta(seconds=20)))
+            curKb.execute("SELECT rechnung, datum FROM buchung WHERE konto = 'FAUKarte' AND (betrag = ? OR betrag = ? OR betrag = ? OR betrag = ? OR betrag = ? OR betrag = ?) AND datum BETWEEN ? AND ? ",(unicode(amount),"{:.0f}".format(amount),"{:.1f}".format(amount), "{:.2f}".format(amount), "{:.3f}".format(amount), "{:.4f}".format(amount), timestamp, timestamp + timedelta(seconds=20)))
```
Do we really need the OR query? For float, `abs(betrag - ?) < 1e-4` should work as well, and for text, we should normalize when saving.

BDrescher replied:
Edit: Thanks for the hint. I did not expect that I could use calculations in the WHERE clause. I have updated the command to your suggestions.

mgmax reviewed (commented):
Not sure whether we really want to change the database, because this will require some conversion work and cause trouble at other places.

BDrescher committed:
Forced the use of Decimal at payment process by asserting for instance Decimal

BDrescher committed:
Updated the payment GUI to use decimals to prevent a wrong display

BDrescher committed:
Add Check of last MagPosLog entry if not booked

BDrescher commented:
Hello,
I've updated the code again and implemented some improvements mentioned by @mgmax .
Additionally I updated the Dialog to use Decimal aswell to prevent false display and added a function to check the last MagPosLog entry if it was not booked. This could be extended further if desired.
As the check is called before every payment, the coverage should be 100%.

mgmax commented on FabLabKasse/UI/FAUcardPaymentDialogCode.py:
```diff
@@ -35,9 +37,11 @@ def __init__(self, parent, amount):
         # Set up member variables and fill GUI
-        self.amount = amount
-        self.label_betrag.setText(u'{:.2f} €'.format(self.amount).replace('.', ','))
+        self.amount = float_to_decimal(amount,2)
```
can't we just pass in amount as Decimal? From a quick look at the code, I'd guess it already is, and we can remove this, (and update `assert isinstance` and `:type amount:` accordingly, and probably remove the import of float_to_decimal)

BDrescher replied:
That is right, we can enforce the usage of Decimal

mgmax commented on FabLabKasse/faucardPayment/FAUcardPaymentThread.py:
```diff
@@ -319,9 +324,9 @@ def check_last_transaction(cur, con):
+            logging.error("CheckTransaction: Kassenterminal vor erfolgreicher Buchung abgestürzt.")
             MagPosLog.save_transaction_result(cur, con, value[1], Decimal(value[2])/100, Info.transaction_ok.value)
-            logging.warning(u"CheckTransaction: Buchung für Karte {0} über Betrag {1} EURCENT fehlt".format(value[1], value[2]) )
+            logging.error(u"CheckTransaction: Buchung für Karte {0} über Betrag {1} EURCENT fehlt".format(value[1], value[2]) )
```
The code is slightly misleading because all three branches of this `if`-`elif`-`else` lead to `return True`, but only the `elif` contains an explicit `return` statement.

mgmax replied:
"Buchung fehlt" -> "Buchung im Kassenbuch fehlt"? Die Meldung ist hier nicht so klar, auch nicht was man denn jetzt tun soll.

BDrescher replied:
Thats correct. The return value should only signal if the check could be performed without issues and does not correlate with the fact if a error was found. I updated the documentation and substituted the `return True` with `pass`

mgmax commented on FabLabKasse/faucardPayment/generateLog.py:
```diff
@@ -144,7 +144,7 @@ def verify_sum(curKb, start, end, magposSum):
 
             # Determine corresponding rechnung
-            curKb.execute("SELECT rechnung, datum FROM buchung WHERE konto = 'FAUKarte' AND (betrag = ? OR betrag = ? OR betrag = ? OR betrag = ? OR betrag = ? OR betrag = ?) AND datum BETWEEN ? AND ? ",(unicode(amount),"{:.0f}".format(amount),"{:.1f}".format(amount), "{:.2f}".format(amount), "{:.3f}".format(amount), "{:.4f}".format(amount), timestamp, timestamp + timedelta(seconds=20)))
+            curKb.execute("SELECT rechnung, datum FROM buchung WHERE konto = 'FAUKarte' AND (abs(betrag - ?) < 1e-4) AND datum BETWEEN ? AND ? ",(unicode(amount), timestamp, timestamp + timedelta(seconds=20)))
```
It is amusing that this SQL statement really works if `betrag` is  column type `TEXT`. Let's consider it a feature of SQLite :-D

mgmax commented on FabLabKasse/faucardPayment/faucard.py:
```diff
@@ -27,7 +27,7 @@ def __init__(self, parent, amount):
         :type amount: Decimal float
```
The `:type:` docs should be updated to match isinstance (for all changes to `isinstance`)

mgmax commented on FabLabKasse/faucardPayment/MagPosLog.py:
```diff
@@ -8,6 +8,8 @@
 
 from faucardStates import Status, Info
+import logging
+from ..shopping.backend.abstract import float_to_decimal
```
unused import?

BDrescher replied:
Forgot to remove after usage was removed before commiting

mgmax reviewed (changes requested):
As far as I understand it, it looks good, with some minor things to fix.

BDrescher committed:
FIX Issue #148, hardcoded DB name in finishLog

BDrescher committed:
Updated documentations, removed unsused imports, forced amount to be given as Decimal at Dialog, documented check_last_transaction of FAUcardPaymentThread better

mgmax commented:
Please feel free to merge.

As a side note: I just pushed commit 7b0c1d2 to development which enforces Decimal for the common payment API (not FAU Card specific).

BDrescher committed:
Added possibility to skip bookings that were booked manually to fix entries
